### PR TITLE
Add support for snapshots

### PR DESCRIPTION
**Pull request by adejanovski**

### CODE REVIEW

File: src/server/src/main/java/io/cassandrareaper/jmx/JmxConnectionFactory.java [hunk 1]
```diff
@@ -118,7 +118,7 @@ protected JmxProxy connect(
     }
   }
 
-  public final JmxProxy connect(Node node, int connectionTimeout)
+  public JmxProxy connect(Node node, int connectionTimeout)
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/jmx/JmxConnectionFactory.java:121]: why unfinalise the method?

(why not mock the method instead? as is the new trend in this project's tests…)

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java [hunk 2]
```diff
@@ -189,9 +189,9 @@ public RepairRun registerRepairRun(
   static int computeGlobalSegmentCount(
       int segmentCountPerNode,
       Map<String, List<RingRange>> endpointToRange) {
-    Preconditions.checkArgument(1 <= endpointToRange.keySet().size());
+    //Preconditions.checkArgument(1 <= endpointToRange.keySet().size());
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java]: why is this now ok?

File: src/server/src/main/java/io/cassandrareaper/service/SnapshotManager.java [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+  public static final String SNAPSHOT_PREFIX = "reaper";
+  private static final Logger LOG = LoggerFactory.getLogger(SnapshotManager.class);
+  private final AppContext context;
+  private final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd-HH-mm-ss");
+  private final ExecutorService executor = Executors.newFixedThreadPool(5);
+  private final Cache<String, Snapshot> cache =
+      CacheBuilder.newBuilder()
+          .maximumSize(1000) // Taille Max
+          .expireAfterWrite(1, TimeUnit.HOURS) // TTL
+          .build();
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/SnapshotManager.java]: the only benefit i can see of the cache is that it avoids a call to `context.storage.getSnapshot(..)`.

What's the measured optimisation that this provides? 

Line 173 can be replaced with `.parallelStream()` and the calls to `context.storage.getSnapshot(..)` will execute in parallel. I would think this is more effective (and better on heap) than the cache approach.

**Comment by michaelsembwever** [comment]
Looks really solid overall. Threw at you a few small questions.

File: src/server/src/main/java/io/cassandrareaper/service/SnapshotManager.java (hunk 3 above)
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/SnapshotManager.java]: As I think having the cache will matter in large clusters, I've moved the cache to a `weakValues()` one as discussed together.

File: src/server/src/main/resources/db/cassandra/011_snapshots.cql [hunk 4]
```diff
@@ -0,0 +1,20 @@
+--
+-- Upgrade to handle snapshots metadata
+
+
+DROP TABLE IF EXISTS snapshot;
```
[the remark this answers is not in the document]
Comment (adejanovski) [reply src/server/src/main/resources/db/cassandra/011_snapshots.cql]: Right, that's not needed/wanted.

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java (hunk 2 above)
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/RepairRunService.java]: It was just a leftover from an attempt to run integration tests that didn't go through.
I've reverted this change.

File: src/server/src/main/java/io/cassandrareaper/jmx/JmxConnectionFactory.java (hunk 1 above)
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/jmx/JmxConnectionFactory.java:121]: That's actually how many of the tests are built and overriding those methods is necessary.
Due to how the JmxConnectionFactory is injected in the context, I couldn't use a mock instead (as far as I can remember).

**Pull request merged**